src/main/java/com/googlesps/feedon/servlets/DonationOfferServlet.java

```diff
@@ -0,0 +1,115 @@
+import com.google.appengine.api.blobstore.BlobInfo;
+import com.google.appengine.api.blobstore.BlobInfoFactory;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.images.ImagesService;
+import com.google.appengine.api.images.ImagesServiceFactory;
+import com.google.appengine.api.images.ServingUrlOptions;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.gson.Gson;
+
+@WebServlet("/donation-offer")
+public class DonationOfferServlet extends HttpServlet {
+
+    private static final String JSON_CONTENT_TYPE = "application/json";
+    private Gson gson = new Gson();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        
+        // Write code here to query all donation offers from Datastore "Donation" entity 
+
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        // Get the input from the form.
+        String restaurantName = request.getParameter("restaurantName"); 
+        String location = request.getParameter("location");
+        String category = request.getParameter("category");
+        String pickUpTime = request.getParameter("pickUpTime");
+        String quantity = request.getParameter("quantity");
+        String specialInstructions = request.getParameter("specialInstructions");
+        
+        // Image URL from Blobstore & current timestamp
+        String imageURL = getUploadedFileUrl(request, "image");
+        Date timestamp = new Date(); 
+
+        // Create Donation offer entity 
+        Entity donationOfferEntity = new Entity("Donation");
+        donationOfferEntity.setProperty("restaurantName", restaurantName);
+        donationOfferEntity.setProperty("location", location);
+        donationOfferEntity.setProperty("category", category);
+        donationOfferEntity.setProperty("pickUpTime", pickUpTime);
+        donationOfferEntity.setProperty("quantity", quantity);
+        donationOfferEntity.setProperty("specialInstructions", specialInstructions);
+        donationOfferEntity.setProperty("imageURL", imageURL);
+        donationOfferEntity.setProperty("timestamp", timestamp);
+
+        // Upload donation offer to Datastore 
+        datastore.put(donationOfferEntity);
+        
+        response.sendRedirect("/index.html");
+    }
+
+    /** Returns a URL that points to the uploaded file, or null if the user didn't upload a file. */
+  @Nullable private String getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
```

Comment:
Can we break this method down a bit?

Reply:
Great suggestion. I think I can break it into two cascading functions, let me just try to implement that!

Reply:
Actually, this method was originally returning image URL AND altering URL to take care of edge cases. So, I had already broken this method into two separate methods, `getUploadedFileUrl` and `formatURL`. The method is lengthy, but it has a single responsibility, so should we just leave it as is?

Reply:
SG